protocol/consensus/consensus.go

```diff
@@ -0,0 +1,250 @@
+package consensus
+
+import (
+	"fmt"
+
+	"github.com/relab/hotstuff"
+	"github.com/relab/hotstuff/core"
+	"github.com/relab/hotstuff/core/eventloop"
+	"github.com/relab/hotstuff/core/logging"
+	"github.com/relab/hotstuff/modules"
+	"github.com/relab/hotstuff/network/netconfig"
+	"github.com/relab/hotstuff/network/sender"
+	"github.com/relab/hotstuff/protocol/kauri"
+	"github.com/relab/hotstuff/protocol/synchronizer/timeout"
+	"github.com/relab/hotstuff/security/blockchain"
+	"github.com/relab/hotstuff/security/certauth"
+	"github.com/relab/hotstuff/service/clientsrv"
+	"github.com/relab/hotstuff/service/committer"
+)
+
+// Consensus provides a default implementation of the Consensus interface
+// for implementations of the ConsensusImpl interface.
+type Consensus struct {
+	impl           modules.ConsensusRules
+	leaderRotation modules.LeaderRotation
+
+	blockChain    *blockchain.BlockChain
+	committer     *committer.Committer
+	commandCache  *clientsrv.CmdCache
+	sender        *sender.Sender
+	auth          *certauth.CertAuthority
+	configuration *netconfig.Config
+	eventLoop     *eventloop.EventLoop
+	logger        logging.Logger
+	opts          *core.Options
+
+	kauri modules.Kauri
+
+	lastVote hotstuff.View
+	view     hotstuff.View
+}
+
+// New returns a new Consensus instance based on the given Rules implementation.
+func New(
+	impl modules.ConsensusRules,
+	leaderRotation modules.LeaderRotation,
+
+	blockChain *blockchain.BlockChain,
+	committer *committer.Committer,
+	commandCache *clientsrv.CmdCache,
+	sender *sender.Sender,
+	auth *certauth.CertAuthority,
+	configuration *netconfig.Config,
+	eventLoop *eventloop.EventLoop,
+	logger logging.Logger,
+	opts *core.Options,
+) *Consensus {
+	cs := &Consensus{
+		impl:           impl,
+		leaderRotation: leaderRotation,
+
+		blockChain:    blockChain,
+		committer:     committer,
+		commandCache:  commandCache,
+		sender:        sender,
+		auth:          auth,
+		configuration: configuration,
+		eventLoop:     eventLoop,
+		logger:        logger,
+		opts:          opts,
+
+		lastVote: 0,
+	}
+
+	cs.eventLoop.RegisterHandler(hotstuff.ProposeMsg{}, func(event any) {
+		cs.OnPropose(cs.view, event.(hotstuff.ProposeMsg))
```

Comment:
If this is called based on a message, who tells the synchronizer to advance the view?